### PR TITLE
[adobepass] Add support for Sling TV (mso slingtv) (closes yt-dlp/yt-dlp#300

### PR DESCRIPTION
**Pull request by wesnm**
…dlp#300)

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [adding new extractor tutorial](https://github.com/ytdl-org/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/ytdl-org/youtube-dl#youtube-dl-coding-conventions) sections
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [x] Improvement
- [ ] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

Add support for Sling TV MSO to Adobepass. Duplicate of PR submitted to youtube-dl.

Tested with http://stream.nbcolympics.com/2018-winter-olympics-nbcsn-evening-feb-8

### CODE REVIEW

**Comment by pukkandan** [comment]
closes #300

File: yt_dlp/extractor/adobepass.py [hunk 1]
```diff
@@ -1565,6 +1570,60 @@ def extract_redirect_url(html, url=None, fatal=False):
...
+                    provider_login_page_res = self._download_webpage_handle(
+                        urlh.geturl(), video_id, 'Sending first bookend.', query=hidden_data)
+
+                    # log in after being redirected
+                    provider_redirect_post_login_res = post_form(provider_login_page_res, 'Logging in', {
+                        mso_info.get('username_field', 'username'): username,
+                        mso_info.get('password_field', 'password'): password,
+                    })
+
+                    provider_association_redirect, urlh = provider_redirect_post_login_res
```
Comment (pukkandan) [inline yt_dlp/extractor/adobepass.py:1604]:
```suggestion
                    provider_association_redirect, urlh = post_form(provider_login_page_res, 'Logging in', {
                        mso_info['username_field']: username,
                        mso_info['password_field': password,
                    })
```

File: yt_dlp/extractor/adobepass.py [hunk 2]
```diff
@@ -1565,6 +1570,60 @@ def extract_redirect_url(html, url=None, fatal=False):
...
+                    provider_refresh_redirect_url = extract_redirect_url(
+                        provider_association_redirect, url=urlh.geturl())
+
+                    # the redirected URL sends us to the final bookend page
+                    last_bookend_page_res = self._download_webpage_handle(
+                        provider_refresh_redirect_url, video_id,
+                        'Downloading Auth Association Redirect Page (meta refresh)')
+
+                    last_bookend_page, urlh = last_bookend_page_res
+
```
Comment (pukkandan) [inline yt_dlp/extractor/adobepass.py:1616]:
```suggestion
                    last_bookend_page, urlh = self._download_webpage_handle(
                        provider_refresh_redirect_url, video_id,
                        'Downloading Auth Association Redirect Page')
```

File: yt_dlp/extractor/adobepass.py [hunk 3]
```diff
@@ -1565,6 +1570,60 @@ def extract_redirect_url(html, url=None, fatal=False):
...
+                    # window history (?) and count how many times
+                    # the browser has been on a page. To stop back buttons?
+                    # After resubmitting, we are redirected to the login form
+                    #
+                    # after the login form, there is a meta-refresh (bypassing some JS)
+                    # that redirects to another bookend page that once again
+                    # looks at the browser window history... It seems to increment
+                    # by two during the process, so do that here.
+                    #
+                    # after the last bookend, a manual form to finish the login
```
Comment (pukkandan) [inline yt_dlp/extractor/adobepass.py:1587]:
```suggestion
                    # SlingTV  has a meta-refresh based authentication, but also
                    # looks at the tab history to count the number of times the
                    # browser has been on a page
```

Also, remove the other comments. Comments in general should explain "why" and the code self-documents the "what"

**Review by pukkandan — CHANGES_REQUESTED** [review]

File: yt_dlp/extractor/adobepass.py (hunk 2 above)
Comment (wesnm) [reply yt_dlp/extractor/adobepass.py:1613]: Fine by me; I was just copying the existing style.

File: yt_dlp/extractor/adobepass.py [hunk 4]
```diff
@@ -1565,6 +1570,40 @@ def extract_redirect_url(html, url=None, fatal=False):
...
+
+                    provider_login_page_res = self._download_webpage_handle(
+                        urlh.geturl(), video_id, 'Sending first bookend.',
+                        query=hidden_data)
+
+                    provider_association_redirect, urlh = post_form(
+                        provider_login_page_res,
+                        'Logging in', {
+                        mso_info['username_field']: username,
+                        mso_info['password_field': password})
```
Comment (pukkandan) [inline yt_dlp/extractor/adobepass.py:1591]:
```suggestion
                    provider_association_redirect, urlh = post_form(
                        provider_login_page_res,  'Logging in', {
                            mso_info['username_field']: username,
                            mso_info['password_field']: password
                        })
```

File: yt_dlp/extractor/adobepass.py [hunk 5]
```diff
@@ -1565,6 +1570,40 @@ def extract_redirect_url(html, url=None, fatal=False):
                         }), headers={
                             'Content-Type': 'application/x-www-form-urlencoded'
                         })
+                elif mso_id == 'slingtv':
+                    # SlingTV  has a meta-refresh based authentication, but also
```
Comment (pukkandan) [inline yt_dlp/extractor/adobepass.py:1574]:
```suggestion
                    # SlingTV has a meta-refresh based authentication, but also
```

File: yt_dlp/extractor/adobepass.py [hunk 6]
```diff
@@ -1565,6 +1570,40 @@ def extract_redirect_url(html, url=None, fatal=False):
...
+
+                    hidden_data = self._hidden_inputs(first_bookend_page)
+                    hidden_data['history'] = 1
+
+                    provider_login_page_res = self._download_webpage_handle(
+                        urlh.geturl(), video_id, 'Sending first bookend.',
+                        query=hidden_data)
+
+                    provider_association_redirect, urlh = post_form(
+                        provider_login_page_res,  'Logging in', {
```
Comment (pukkandan) [inline yt_dlp/extractor/adobepass.py:1588]:
```suggestion
                        provider_login_page_res, 'Logging in', {
```

**Review by pukkandan — APPROVED** [review]
Please test and confirm that we havent broken anything

**Comment by wesnm** [comment]
Confirmed:

```
[debug] Command-line config: ['--verbose', '--ap-mso', 'slingtv', '--ap-username', 'PRIVATE', '--ap-
password', 'PRIVATE', '--skip-download', 'http://stream.nbcolympics.com/2018-winter-olympics-nbcsn-e
vening-feb-8']                
                              
[debug] Encodings: locale UTF-8, fs utf-8, out utf-8, pref UTF-8
[debug] yt-dlp version 2021.07.24 (source)
[debug] Plugin Extractors: ['SamplePlugin']
[debug] Git HEAD: b8be3bba5   
[debug] Python version 3.9.6 (CPython 64bit) - FreeBSD-13.0-STABLE-amd64-64bit-ELF
[debug] exe versions: ffmpeg 4.4, ffprobe 4.4, rtmpdump 2.4
[debug] Proxy map: {}
[debug] [nbcolympics:stream] Extracting URL: http://stream.nbcolympics.com/2018-winter-olympics-nbcsn-evening-feb-8
[nbcolympics:stream] 2018-winter-olympics-nbcsn-evening-feb-8: Downloading webpage
[nbcolympics:stream] 203493: Downloading JSON metadata
[nbcolympics:stream] 203493: Downloading JSON metadata
[nbcolympics:stream] 203493: Downloading Provider Redirect Page
[nbcolympics:stream] 203493: Sending first bookend.
[nbcolympics:stream] 203493: Logging in
[nbcolympics:stream] 203493: Downloading Auth Association Redirect Page
[nbcolympics:stream] 203493: Sending final bookend.
[nbcolympics:stream] 203493: Confirming Login
[nbcolympics:stream] 203493: Retrieving Session
[nbcolympics:stream] 203493: Retrieving Media Token
[nbcolympics:stream] 203493: Downloading webpage
[nbcolympics:stream] 203493: Downloading m3u8 information
[debug] Formats sorted by: hasvid, ie_pref, lang, quality, res, fps, vcodec:vp9.2(10), acodec, filesize, fs_approx, tbr, vbr, abr, asr, proto, vext, aext, hasaud, source, id
[info] 203493: Downloading 1 format(s): 5726-1
```

**Pull request merged**